### DISCUSSION TITLE
How do I get POST data from http request?

### DISCUSSION THREAD

**Discussion by suchislife801** [denoland/deno#9353] [Q&A] [answered]
Hello. I can't seem to find this information anywhere.

If I make a post request with JSON data, how do I get this posted data from the request?

```
const reader = new Deno.Buffer(httpRequest.r.buf as ArrayBuffer);
const bufferContent = Deno.readAllSync(reader);

console.log(bufferContent);
```

TS2341 [ERROR]: Property 'buf' is private and only accessible within class 'BufReader'

**Comment by satyarohith** [comment, marked as answer]
You can access the data using `.body` property of `ServerRequest` ([docs](https://doc.deno.land/https/deno.land/std/http/mod.ts#ServerRequest)).

Here's a small example:
```ts
import { serve } from "https://deno.land/std@0.85.0/http/server.ts";
const s = serve({ port: 8000 });
console.log("http://localhost:8000/");
const decoder = new TextDecoder();
for await (const req of s) {
  if (req.method === "POST") {
    const buf = await Deno.readAll(req.body);
    const json = JSON.parse(decoder.decode(buf));
    console.log(json);
  }
  req.respond({ body: "Hello World\n" });
}
```
Run the above program and post some data to the endpoint to see the posted data logged to the console.
```sh
curl -x POST -d '{"name": "Deno"}' http://localhost:8000
```

**Reply by brandonros** [reply]
Can you show how to do this with a non-text request body (like application/octect-stream)?

**Reply by kitsonk** [reply]
The `buf` in the example is a `Uint8Array`.  That contains the binary data which you would do with as you please.
